core/src/main/java/org/apache/iceberg/deletes/RoaringPositionBitmap.java

```diff
@@ -0,0 +1,309 @@
+   */
+  public void forEach(LongConsumer consumer) {
+    for (int index = 0; index < bitmaps.length; index++) {
+      forEach(bitmaps[index], index, consumer);
+    }
+  }
+
+  /**
+   * Returns the number of bytes required to serialize the bitmap.
+   *
+   * @return the serialized size in bytes
+   */
+  public long serializedSizeInBytes() {
+    long size = BITMAP_COUNT_SIZE_BYTES;
+    for (RoaringBitmap bitmap : bitmaps) {
+      size += BITMAP_KEY_SIZE_BYTES + bitmap.serializedSizeInBytes();
+    }
+    return size;
+  }
+
+  /**
+   * Serializes the bitmap using the portable serialization format described below.
+   *
+   * <ul>
+   *   <li>The number of 32-bit Roaring bitmaps, serialized as 8 bytes
+   *   <li>For each 32-bit Roaring bitmap, ordered by unsigned comparison of the 32-bit keys:
+   *       <ul>
+   *         <li>The key stored as 4 bytes
+   *         <li>Serialized 32-bit Roaring bitmap using the standard format
+   *       </ul>
+   * </ul>
+   *
+   * <p>Note the byte order of the buffer must be little-endian.
+   *
+   * @param buffer the buffer to write to
+   * @see <a href="https://github.com/RoaringBitmap/RoaringFormatSpe">Roaring bitmap spec</a>
+   */
+  public void serialize(ByteBuffer buffer) {
+    validateByteOrder(buffer);
+    buffer.putLong(bitmaps.length);
+    for (int index = 0; index < bitmaps.length; index++) {
+      buffer.putInt(index);
+      bitmaps[index].serialize(buffer);
+    }
+  }
+
+  private void allocateBitmapsIfNeeded(int requiredLength) {
+    if (bitmaps.length < requiredLength) {
+      if (bitmaps.length == 0 && requiredLength == 1) {
+        this.bitmaps = new RoaringBitmap[] {new RoaringBitmap()};
+      } else {
+        RoaringBitmap[] newBitmaps = new RoaringBitmap[requiredLength];
+        System.arraycopy(bitmaps, 0, newBitmaps, 0, bitmaps.length);
+        for (int index = bitmaps.length; index < requiredLength; index++) {
+          newBitmaps[index] = new RoaringBitmap();
+        }
+        this.bitmaps = newBitmaps;
+      }
+    }
+  }
+
+  /**
+   * Deserializes a bitmap from a buffer, assuming the portable serialization format.
+   *
+   * @param buffer the buffer to read from
+   * @return a new bitmap instance with the deserialized data
+   */
+  public static RoaringPositionBitmap deserialize(ByteBuffer buffer) {
+    validateByteOrder(buffer);
+
+    // the bitmap array may be sparse with more elements than the number of read bitmaps
+    int remainingBitmapCount = readBitmapCount(buffer);
+    List<RoaringBitmap> bitmaps = Lists.newArrayListWithExpectedSize(remainingBitmapCount);
+    int lastKey = 0;
+
+    while (remainingBitmapCount > 0) {
+      int key = readKey(buffer, lastKey);
+
+      // fill gaps as the bitmap array may be sparse
+      while (lastKey < key) {
+        bitmaps.add(new RoaringBitmap());
+        lastKey++;
+      }
+
+      RoaringBitmap bitmap = readBitmap(buffer);
+      bitmaps.add(bitmap);
+
+      lastKey++;
```

Comment:
I think this is correct, but it's a little confusing.

When the `while (lastKey < key)` loop ends, `lastKey` is going to be _equal to_ `key`. That seems correct to me given that `lastKey` is supposed to be the previous iteration's value of `key`.

However, this increments `lastKey` so it is actually the previous iteration's `key + 1`. That's what is strange to me. I think this is also why `readKey` checks that `key >= lastKey` because it actually will be equal if the keys are incrementing.

I'd probably update this by:
* Setting `lastKey` to `key` here
* Changing the check in `readKey` to `lastKey < key`
* Updating the `while` loop to `while (lastKey < key - 1)`

Reply:
Updated.